lib/HTFeed/JobMetrics.pm

```diff
@@ -0,0 +1,320 @@
+	$self->{file} = $ENV{'HTFEED_JOBMETRICS_DATA_DIR'};
+    }
+    $self->{prom} = Prometheus::Tiny::Shared->new(
+	filename => $self->{file}
+    );
+
+    bless($self, $class);
+    $self->_setup_metrics();
+
+    return $self;
+}
+
+# Show where data is stored.
+sub loc {
+    my $self = shift;
+    $self->{file};
+}
+
+# for counters, where each observation is an arbitrary +=,
+# e.g. $jm->add("Download_ms", 219.5);
+# or $ perl lib/HTFeed/JobMetrics.pm --add Download_ms 219.5
+sub add {
+    my $self   = shift;
+    my $metric = shift;
+    my $value  = shift;
+
+    # Make sure $value is numeric
+    unless ($value + 0) {
+	get_logger()->warn("Non-numeric value \"$value\" given for $metric");
+	return;
+    }
+
+    $self->_valid_metric($metric) && $self->{prom}->add($metric, $value);
+}
+
+# for integer counters, where each observation is a single tick ++
+# e.g. $jm->inc("Unpack_items");
+# or $ perl JobMetrics.pm --inc Unpack_items
+sub inc {
+    my $self   = shift;
+    my $metric = shift;
+
+    $self->_valid_metric($metric) && $self->{prom}->inc($metric);
+}
+
+# for histograms
+# e.g. $jm->observe("Unpack_ms", 125);
+# or $ perl JobMetrics.pm --observe Unpack_ms 125
+# Note that histograms must be set up with buckets ahead of time.
+sub observe {
+    my $self   = shift;
+    my $metric = shift;
+    my $value  = shift;
+
+    # Make sure $value is numeric
+    unless ($value + 0) {
+	get_logger()->warn("Non-numeric value \"$value\" given for $metric");
+	return;
+    }
+
+    $self->_valid_metric($metric) && $self->{prom}->histogram_observe($metric, $value);
+}
+
+# just show names of metrics
+sub list_metrics {
+    my $self = shift;
+    [sort keys %{$self->{metrics}}];
+}
+
+# pretty-printable tabular data
+sub pretty {
+    my $self = shift;
+    $self->{prom}->format;
+}
+
+# Return metrics matching $pat
+sub match {
+    my $self = shift;
+    my $pat  = shift;
+    my @pretty = split("\n", $self->pretty);
+    [grep { /$pat/ } @pretty];
+}
+
+# irrevocably delete job metrics
+sub clear {
+    my $self = shift;
+    $self->{prom}->clear;
+}
+
+# "private" from here on, indicated with _leading_underscores
+# in method names
+
+# Check that only valid metrics are used,
+# or make an intelligent complaint.
+# e.g. $jm->_valid_metric("Unpack_ms") # -> 1
+# e.g. $jm->_valid_metric("ms pacman") # -> warn & 0
+sub _valid_metric {
+    my $self   = shift;
+    my $metric = shift;
+
+    return 1 if defined $self->{metrics}->{$metric};
+
+    # caller 1: we don't want the line number from the call inside this package.
+    # remember, this is a "private" method.
+    my (undef, $filename, $line) = caller(1);
+    get_logger()->warn("invalid metric name \"$metric\" at $filename:$line");
+    return 0;
+}
+
+sub _setup_metrics {
+    my $self = shift;
+    my $prom = $self->{prom};
+
+    # Each element x in this array gets turned into 3 prom declarations:
+    # x_ms, x_bytes and x_items
+    my @measurable_stages = (
+	'Collate',
```

Comment:
To better fit the naming scheme for prometheus metrics I might suggest lowercasing these and prepending `ingest_` or `ingest_stage_`

Reply:
New naming scheme uses `ingest_` prefix and lowercases the stage name. What used to be `Pack_ms` is now, after all the changes, `ingest_pack_seconds`.